# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [46]:
import os
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
%matplotlib inline
pd.options.display.float_format = '{:.2f}'.format #breaking out of scientific notations for the full notebook
os.getcwd()

'C:\\Users\\Vatsla Issar\\Downloads'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [1]:
df=pd.read_csv('step3_output.csv')
df.head(3)

<IPython.core.display.Javascript object>

Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   

   vertical_drop  base_elev  trams  fastEight  fastSixes  fastQuads  ...  \
0            294       1796      0        0.0          0          0  ...   
1           1800       9200      0        0.0          0          1  ...   
2            600       7200      0        0.0          0          0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000             150.0       36.0   
1              800.0       80.000000             115.0       49.0   
2               88.0      174.873239             110.0       84.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0          34.0              152.0   
1            250.0          74.0          78.0              104.0   
2            300.0          47.0          47.0              107.0   

   NightSkiing_ac  clusters  
0       30.000000         0  
1       80.000000         2  
2      100.395722         1  

[3 rows x 28 columns]

In [2]:
df=df.drop(['Unnamed: 0'], axis=1)
df.head(3)

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   

   base_elev  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0       1796      0        0.0          0          0     0  ...   
1       9200      0        0.0          0          1     2  ...   
2       7200      0        0.0          0          0     0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000             150.0       36.0   
1              800.0       80.000000             115.0       49.0   
2               88.0      174.873239             110.0       84.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0          34.0              152.0   
1            250.0          74.0          78.0              104.0   
2            300.0          47.0          47.0              107.0   

   NightSkiing_ac  clusters  
0       30.000000         0  
1       80.000000         2  
2      100.395722         1  

[3 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
state_d=pd.get_dummies(df['state'])
state_d.head()

<IPython.core.display.Javascript object>

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0       1        0           0         0            0      0         0   
1       0        1           0         0            0      0         0   
2       0        0           1         0            0      0         0   
3       0        0           1         0            0      0         0   
4       0        0           1         0            0      0         0   

   Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0        0     0      0  ...             0             0          0     0   
1        0     0      0  ...             0             0          0     0   
2        0     0      0  ...             0             0          0     0   
3        0     0      0  ...             0             0          0     0   
4        0     0      0  ...             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0  

[5 rows x 35 columns]

In [4]:
df = pd.concat([df, state_d], axis=1)

<IPython.core.display.Javascript object>

In [5]:
df=df.drop('state', axis=1)

In [6]:
df.head().T #state dropped and state_d (dummies of state) are included.

0                    1  \
Name           Hilltop Ski Area  Sunrise Park Resort   
summit_elev                2090                11100   
vertical_drop               294                 1800   
base_elev                  1796                 9200   
trams                         0                    0   
...                         ...                  ...   
Virginia                      0                    0   
Washington                    0                    0   
West Virginia                 0                    0   
Wisconsin                     0                    0   
Wyoming                       0                    0   

                                           2                 3              4  
Name           Yosemite Ski & Snowboard Area  Donner Ski Ranch  June Mountain  
summit_elev                             7800              8012          10090  
vertical_drop                            600               750           2590  
base_elev                               7200              7031           7545  
trams                                      0                 0              0  
...                                      ...               ...            ...  
Virginia                                   0                 0              0  
Washington                                 0                 0              0  
West Virginia                              0                 0              0  
Wisconsin                                  0                 0              0  
Wyoming                                    0                 0              0  

[61 rows x 5 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [8]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=lm.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [13]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test,y_pred))

-1.7994782180661886e+24
4134761523721.643


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [14]:
lm.intercept_

1052484751547.9099

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [15]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
a=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

<IPython.core.display.Javascript object>

In [16]:
sort_a=a.sort_values(by='Coefficient', ascending=False)
sort_a

Coefficient
fastSixes          1.061488e+14
fastEight          7.815730e+13
New York           2.633478e+13
Michigan           2.490380e+13
total_chairs       2.183589e+13
New Hampshire      2.160882e+13
trams              2.072264e+13
Wisconsin          2.066998e+13
Pennsylvania       2.066998e+13
Minnesota          1.860669e+13
California         1.860669e+13
Vermont            1.746031e+13
Idaho              1.621618e+13
New Mexico         1.621618e+13
Wyoming            1.484976e+13
Colorado           1.484976e+13
Maine              1.484976e+13
Connecticut        1.484976e+13
Ohio               1.332347e+13
Washington         1.332347e+13
Massachusetts      1.332347e+13
surface            1.249475e+13
double             1.201021e+13
Iowa               1.157424e+13
Montana            1.157424e+13
Oregon             1.157424e+13
Illinois           1.157424e+13
Utah               1.157424e+13
West Virginia      1.157424e+13
North Carolina     1.157424e+13
triple             1.034243e+13
Indiana            9.479449e+12
Missouri           9.479449e+12
Virginia           9.479449e+12
Nevada             6.723513e+12
Arizona            6.723513e+12
South Dakota       6.723513e+12
Alaska             6.723513e+12
Rhode Island       6.723513e+12
Tennessee          6.723513e+12
quad               6.235435e+12
fastQuads          4.968913e+12
Maryland           2.918495e+10
New Jersey         2.171117e+09
summit_elev        5.265014e+01
base_elev          5.169208e+01
AdultWeekday       1.047461e+01
vertical_drop      9.322146e+00
clusters           4.242188e+00
projectedDaysOpen  1.246094e+00
daysOpenLastYear   1.193359e+00
Runs               1.148438e+00
Snow Making_ac     9.870605e-01
LongestRun_mi      9.179688e-01
averageSnowfall    6.347656e-01
TerrainParks       5.561523e-01
yearsOpen          7.812500e-02
NightSkiing_ac     6.835938e-02
SkiableTerrain_ac  1.757812e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [17]:
df2=pd.read_csv('step3_output.csv') #Loaded again because we created dummies for state earlier (multiple levels)
df2=df2.drop(['Unnamed: 0'], axis=1) #df=df.drop(state_d,axis=1) would have worked but i wanted to keep it seperate
df2.head().T

<IPython.core.display.Javascript object>

0                    1  \
Name               Hilltop Ski Area  Sunrise Park Resort   
state                        Alaska              Arizona   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
base_elev                      1796                 9200   
trams                             0                    0   
fastEight                         0                    0   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                             13                   65   
TerrainParks                      1                    2   
LongestRun_mi                     1                  1.2   
SkiableTerrain_ac                30                  800   
Snow Making_ac                   30                   80   
daysOpenLastYear                150                  115   
yearsOpen                        36                   49   
averageSnowfall                  69                  250   
AdultWeekday                     30                   74   
AdultWeekend                     34                   78   
projectedDaysOpen               152                  104   
NightSkiing_ac                   30                   80   
clusters                          0                    2   

                                               2                 3  \
Name               Yosemite Ski & Snowboard Area  Donner Ski Ranch   
state                                 California        California   
summit_elev                                 7800              8012   
vertical_drop                                600               750   
base_elev                                   7200              7031   
trams                                          0                 0   
fastEight                                      0                 0   
fastSixes                                      0                 0   
fastQuads                                      0                 0   
quad                                           0                 0   
triple                                         1                 1   
double                                         3                 5   
surface                                        1                 2   
total_chairs                                   5                 8   
Runs                                          10                52   
TerrainParks                                   2                 2   
LongestRun_mi                                0.4               1.5   
SkiableTerrain_ac                             88               505   
Snow Making_ac                           174.873                60   
daysOpenLastYear                             110               163   
yearsOpen                                     84                82   
averageSnowfall                              300               400   
AdultWeekday                                  47                75   
AdultWeekend                                  47                75   
projectedDaysOpen                            107               170   
NightSkiing_ac                           100.396           100.396   
clusters                                       1                 1   

                               4  
Name               June Mountain  
state                 California  
summit_elev                10090  
vertical_drop               2590  
base_elev                   7545  
trams                          0  
fastEight                      0  
fastSixes                      0  
fastQuads                      2  
quad                           0  
triple 

In [18]:
df2=df2.drop('state', axis=1)
df2.head().T

0                    1  \
Name               Hilltop Ski Area  Sunrise Park Resort   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
base_elev                      1796                 9200   
trams                             0                    0   
fastEight                         0                    0   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                             13                   65   
TerrainParks                      1                    2   
LongestRun_mi                     1                  1.2   
SkiableTerrain_ac                30                  800   
Snow Making_ac                   30                   80   
daysOpenLastYear                150                  115   
yearsOpen                        36                   49   
averageSnowfall                  69                  250   
AdultWeekday                     30                   74   
AdultWeekend                     34                   78   
projectedDaysOpen               152                  104   
NightSkiing_ac                   30                   80   
clusters                          0                    2   

                                               2                 3  \
Name               Yosemite Ski & Snowboard Area  Donner Ski Ranch   
summit_elev                                 7800              8012   
vertical_drop                                600               750   
base_elev                                   7200              7031   
trams                                          0                 0   
fastEight                                      0                 0   
fastSixes                                      0                 0   
fastQuads                                      0                 0   
quad                                           0                 0   
triple                                         1                 1   
double                                         3                 5   
surface                                        1                 2   
total_chairs                                   5                 8   
Runs                                          10                52   
TerrainParks                                   2                 2   
LongestRun_mi                                0.4               1.5   
SkiableTerrain_ac                             88               505   
Snow Making_ac                           174.873                60   
daysOpenLastYear                             110               163   
yearsOpen                                     84                82   
averageSnowfall                              300               400   
AdultWeekday                                  47                75   
AdultWeekend                                  47                75   
projectedDaysOpen                            107               170   
NightSkiing_ac                           100.396           100.396   
clusters                                       1                 1   

                               4  
Name               June Mountain  
summit_elev                10090  
vertical_drop               2590  
base_elev                   7545  
trams                          0  
fastEight                      0  
fastSixes                      0  
fastQuads                      2  
quad                           0  
triple                         0  
double                         4  
surface                        1  
total_chairs                   7  
Runs                          41

In [19]:
X2 = df2.drop(['Name','AdultWeekend'], axis=1)
y2 = df2['AdultWeekend'] 

In [20]:
scaler2 = preprocessing.StandardScaler().fit(X2)

In [21]:
X_scaled2=scaler2.transform(X2)

In [22]:
y2 = y2.ravel()

X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)

In [23]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train2,y_train2)

In [24]:
y_pred2=lm2.predict(X_test2)

In [25]:
print(explained_variance_score(y_test2, y_pred2))
print(mean_absolute_error(y_test2,y_pred2))

0.5502055140054993
7.120499596523722


In [26]:
lm2.intercept_

56.31479302198481

In [27]:
b=pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])
b.sort_values(by='Coefficient', ascending=False)

<IPython.core.display.Javascript object>

Coefficient
AdultWeekday       1.179109e+01
base_elev          5.571708e+00
summit_elev        4.417999e+00
daysOpenLastYear   2.518922e+00
vertical_drop      1.559254e+00
projectedDaysOpen  1.494292e+00
averageSnowfall    1.424605e+00
surface            1.325193e+00
quad               1.324325e+00
double             8.456285e-01
clusters           8.452888e-01
SkiableTerrain_ac  7.251321e-01
TerrainParks       6.926718e-01
Snow Making_ac     5.902934e-01
triple             4.823348e-01
total_chairs       2.961923e-01
yearsOpen          2.533168e-01
Runs               2.000474e-01
fastQuads          7.584184e-02
LongestRun_mi      6.790844e-02
NightSkiing_ac     4.194331e-02
fastEight          2.420286e-14
fastSixes          2.997602e-15
trams              2.664535e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [29]:
df3=df2.drop(['summit_elev','base_elev'],axis=1)

In [30]:
df3.columns

Index(['Name', 'vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads',
       'quad', 'triple', 'double', 'surface', 'total_chairs', 'Runs',
       'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [31]:
X3 = df3.drop(['Name','AdultWeekend'], axis=1)
y3 = df3['AdultWeekend'] 

In [32]:
scaler3 = preprocessing.StandardScaler().fit(X3)

In [33]:
X_scaled3=scaler3.transform(X3)

In [34]:
y3 = y3.ravel()

X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled3, y3, test_size=0.25, random_state=1)

In [35]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train3,y_train3)
y_pred3=lm3.predict(X_test3)
print(explained_variance_score(y_test3, y_pred3))
print(mean_absolute_error(y_test3,y_pred3))
print(lm3.intercept_)

0.557322419344288
7.071656285426007
56.290548239095735


In [36]:
c=pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])
c.sort_values(by='Coefficient', ascending=False)

<IPython.core.display.Javascript object>

Coefficient
AdultWeekday       1.178102e+01
vertical_drop      2.462473e+00
daysOpenLastYear   2.392467e+00
clusters           2.223592e+00
averageSnowfall    1.599924e+00
surface            1.326846e+00
projectedDaysOpen  1.302945e+00
quad               1.270529e+00
double             8.793738e-01
SkiableTerrain_ac  7.952728e-01
TerrainParks       7.190345e-01
triple             4.679973e-01
Snow Making_ac     4.454458e-01
total_chairs       3.210833e-01
yearsOpen          1.934435e-01
Runs               8.931853e-02
fastQuads          5.348338e-02
NightSkiing_ac     1.456175e-02
LongestRun_mi      7.259269e-03
fastSixes          8.881784e-16
trams              4.440892e-16
fastEight          1.110223e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.7994782180661886e+24 | 4134761523721.643 |-|
| Model 2. | 0.55021| 7.120499 |'state'|
| Model 3. | 0.557322 | 7.071656 |'state','summit_elev','base_elev'|

Model Selection:

In [ ]:
#Model 3 explains the most amount of variance and the MAE is the least for model 3. So Model 3 should be selected.